File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """

        fast = head
        slow = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        tail = None
        current = slow.next
        slow.next = None

        while current:
            temp = current.next
            current.next = tail
            tail = current
            current = temp

        l1 = head
        l2 = tail 

        while l1 and l2:
            temp = l1.next
            l1.next = l2
            l1 = l1.next
            l2 = temp
        

File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }

        ListNode tail = null;
        ListNode current = slow.next;
        slow.next = null;

        while(current != null){
            ListNode temp = current.next;
            current.next = tail;
            tail = current;
            current = temp;
        }
        
        ListNode l1 = head;
        ListNode l2 = tail;

        while(l1 != null && l2 != null){
            ListNode temp = l1.next;
            l1.next = l2;
            l1 = l1.next;            
            l2 = temp;
        }
    }
}